### ISSUE TITLE
pgroonga: failed to remove: <Sources152012>: system call error: No such file or directory: failed to stat. while cleaning up index of relation 2PC issues

### ISSUE THREAD

**Issue by rwasef1830** [pgroonga/pgroonga#252]
Hello,
I am using pgroonga in a project, and I'm facing a serious issue related to 2 phase commit. Or mix use of tables in 2 phase commit and normal commit mode when one transaction rolls back. When I call VACUUM ANALYZE in the end I get this error.

I don't know the exact steps to reproduce it because it seems to happen randomly, but it's closely related to 2 phase commit (successful and roll backs of prepared statements). If I stop using prepared statements, the issue does not occur. It is solved by doing REINDEX DATABASE, but until then, operations fail randomly with this error since it seems pgroonga/groonga state became out of sync or corrupted somehow.

Using pgroonga "safer" mode hides the issue but at the cost of extremely lower insert performance.

**Comment by rwasef1830** [comment]
In order to assist with more info. I have attached pgroonga log from many months to help with any analysis.
[pgroonga.zip](https://github.com/pgroonga/pgroonga/files/9997433/pgroonga.zip)

**Comment by rwasef1830** [comment]
I see in the log, that the Sources152012 object was created and then removed. I think pgroonga is getting corrupted by transaction rollback or 2 phase transaction rollback.

**Comment by komainu8** [comment]
Thank you for your report!
I'm confirming this problem.

**Comment by rwasef1830** [comment]
More context: The error is triggered at the end of a large bulk insert operation done with an ORM. I am seeing in the log that creating and removing objects in pgroonga seems related to DDL statements. My ORM maybe performing the bulk insertions using a temporary table or something like that. I will try to get a log statement of the queries the ORM runs.

**Comment by rwasef1830** [comment]
I will try to set pgroonga log level to max and try to recreate the problem

**Comment by rwasef1830** [comment]
More info discovered, part of the script where this happens also creates a table and creates multiple groonga index for this table on different columns (text and json), and this is inside a prepared statement, which is later commit at the end, or rolled back in case of error.

I'm trying to reproduce the problem with a script like this:
```sql
DROP TABLE IF EXISTS t2;
DROP TABLE IF EXISTS t3;
SET pgroonga.log_level = dump;

ROLLBACK PREPARED 'test';

BEGIN TRANSACTION;
create table if not exists t3 (text text default ''::text not null);
create index if not exists t3_data_index on t3 using pgroonga (text);
INSERT INTO t3 (text) VALUES (gen_random_uuid());
INSERT INTO t3 (text) VALUES (gen_random_uuid());
INSERT INTO t3 (text) VALUES (gen_random_uuid());
INSERT INTO t3 (text) VALUES (gen_random_uuid());
INSERT INTO t3 (text) VALUES (gen_random_uuid());
PREPARE TRANSACTION 'test';

BEGIN TRANSACTION;
create table if not exists t2 (text text default ''::text not null);
create index if not exists t2_data_index on t2 using pgroonga (text);
INSERT INTO t2 (text) VALUES (gen_random_uuid());
INSERT INTO t2 (text) VALUES (gen_random_uuid());
INSERT INTO t2 (text) VALUES (gen_random_uuid());
INSERT INTO t2 (text) VALUES (gen_random_uuid());
INSERT INTO t2 (text) VALUES (gen_random_uuid());
PREPARE TRANSACTION 'test2';

ROLLBACK PREPARED 'test2';
-- ROLLBACK PREPARED 'test';

BEGIN TRANSACTION;
create table if not exists t2 (text text default ''::text not null);
create index if not exists t2_data_index on t2 using pgroonga (text);
INSERT INTO t2 (text) VALUES (gen_random_uuid());
PREPARE TRANSACTION 'test3';

VACUUM ANALYZE;
COMMIT PREPARED 'test3';

```

But so far, trying to search for the bug, it stopped appearing when intentionally repeating the steps again.

**Comment by rwasef1830** [comment]
@kou SUCCESS in repro with this script:

```
DROP TABLE IF EXISTS t2;
DROP TABLE IF EXISTS t3;
SET pgroonga.log_level = dump;

ROLLBACK PREPARED 'test';

BEGIN TRANSACTION;
create table if not exists t3 (text text default ''::text not null, json jsonb default '{}'::jsonb not null);
create index if not exists t3_data_index on t3 using pgroonga (text);
create index if not exists t3_data_index2 on t3 using pgroonga (json);
INSERT INTO t3 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t3 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t3 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t3 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t3 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
PREPARE TRANSACTION 'test';

BEGIN TRANSACTION;
create table if not exists t2 (text text default ''::text not null, json jsonb default '{}'::jsonb not null);
create index if not exists t2_data_index on t2 using pgroonga (text);
create index if not exists t2_data_index2 on t2 using pgroonga (json);
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
PREPARE TRANSACTION 'test2';

ROLLBACK PREPARED 'test2';
-- ROLLBACK PREPARED 'test';

BEGIN TRANSACTION;
create table if not exists t2 (text text default ''::text not null, json jsonb default '{}'::jsonb not null);
create index if not exists t2_data_index on t2 using pgroonga (text);
create index if not exists t2_data_index2 on t2 using pgroonga (json);
INSERT INTO t2 (text, json) VALUES (gen_random_uuid(), jsonb_build_object(gen_random_uuid(), gen_random_uuid()));
PREPARE TRANSACTION 'test3';

VACUUM ANALYZE;
COMMIT PREPARED 'test3';
VACUUM ANALYZE;
```

Result:
ERROR:  22023: pgroonga: object isn't found: <Sources2985924>
CONTEXT:  while cleaning up index "t2_data_index" of relation "public.t2"
LOCATION:  PGrnCheckRC, pgrn-groonga.c:168



The issue seems to be related that groonga by mistake deletes some entries from wal in vaccuum and doesn't take into account they are involved in prepared statement.

**Comment by rwasef1830** [comment]
@komainu8 Actually, I see this issue can happen to any database using 2PC by the autovacuum if it happened by chance at the moment that there is any prepared transaction involving groonga index columns, so the database will eventually self-corrupt.

**Comment by komainu8** [comment]
Thank you for providing reproduce procedure!
However, I can't reproduce in my envirment.

Probably, the cause of problem VACUUM is executed from multiple processes in the same time.
We'll try implement a exclusion control of VACUUM.

**Comment by rwasef1830** [comment]
@komainu8 In my case I am using postgres 14.5 with pgroonga 2.3.9, windows, what environment are you testing on ?
Did you try the latest script I put above ? With running it with psql on blank database and enabled prepared transaction on postgres, it reproduces the issue immediately.

I will test newest pgroonga and newest postgres and see if the issue still remains there.

**Comment by komainu8** [comment]
Thank you for  providing reproduce procedure again!

I'll try reproduce in my environment again. (I use PostgreSQL 15.0 and PGroonga 2.4.1 in Debian GNU/Linux. If this problem don't reproduce in Linux, I try reproduce in Windows too.)

**Comment by komainu8** [comment]
I can reproduce this problem with your script!
I'll go ahead with debug.

**Comment by rwasef1830** [comment]
@komainu8 is there any chance this will be fixed in 2.4.2?  have you found the cause of the problem?
Is it a minor bug? or a major issue that requires a large refactor or rearchitect to solve?

**Comment by kou** [comment]
I've fixed this.
It will be included in 2.4.2.
